src/Microsoft.AspNetCore.Server.IISIntegration/Server/IO/AsyncIOEngine.cs

```diff
@@ -0,0 +1,171 @@
+        }
+
+        public ValueTask<int> WriteAsync(ReadOnlySequence<byte> data)
+        {
+            var write = GetWriteOperation();
+            write.Initialize(_handler, data);
+            Run(write);
+            return new ValueTask<int>(write, 0);
+        }
+
+        private void Run(AsyncIOOperation ioOperation)
+        {
+            lock (this)
+            {
+                if (_stopped)
+                {
+                    throw new IOException("IO stopped", NativeMethods.ERROR_OPERATION_ABORTED);
+                }
+
+                if (_runningOperation != null)
+                {
+                    if (_nextOperation == null)
+                    {
+                        _nextOperation = ioOperation;
+
+                        if (_runningOperation is AsyncReadOperation)
+                        {
+                            NativeMethods.HttpTryCancelIO(_handler);
+                        }
+                    }
+                    else
+                    {
+                        throw new InvalidOperationException("Only one queued operation is allowed");
+                    }
+                }
+                else
+                {
+                    // we are just starting operation so there would be no
+                    // continuation registered
+                    var completed = ioOperation.Invoke() != null;
+
+                    // operation went async
+                    if (!completed)
+                    {
+                        _runningOperation = ioOperation;
+                    }
+                }
+            }
+        }
+
+
+        public ValueTask FlushAsync()
+        {
+            var flush = GetFlushOperation();
+            flush.Initialize(_handler);
+            Run(flush);
+            return new ValueTask(flush, 0);
+        }
+
+        public void NotifyCompletion(int hr, int bytes)
+        {
+            AsyncIOOperation.AsyncContinuation continuation;
+            AsyncIOOperation.AsyncContinuation? nextContinuation = null;
+
+            lock (this)
+            {
+                Debug.Assert(_runningOperation != null);
+
+                continuation = _runningOperation.Complete(hr, bytes);
+
+                var next = _nextOperation;
+                _nextOperation = null;
+                _runningOperation = null;
+
+                if (next != null)
+                {
+                    if (_stopped)
+                    {
+                        // Abort next operation if IO is stopped
+                        nextContinuation = next.Complete(NativeMethods.ERROR_OPERATION_ABORTED, 0);
+                    }
+                    else
+                    {
+                        nextContinuation = next.Invoke();
+
+                        // operation went async
+                        if (nextContinuation == null)
+                        {
+                            _runningOperation = next;
+                        }
```

Comment:
Do we need to null out _nextOperation if it completed synchronously?

Reply:
We assign both to null couple lines above.

Reply:
oh im blind.